core/src/main/java/org/radargun/config/ClasspathScanner.java

```diff
@@ -0,0 +1,103 @@
+         } else if (resourceFile.isDirectory()) {
+            int prefixLength = resource.toString().length() + 1;
+            try {
+               Files.find(resourceFile.toPath(), Integer.MAX_VALUE, (path, attrs) -> attrs.isRegularFile() && (path.toString().endsWith(".class") || path.toString().endsWith(".jar"))).forEach(file -> {
+                  scanFile(file.toString(), prefixLength, superClass, annotationClass, requirePackage, consumer);
+               });
+            } catch (IOException e) {
+               throw new RuntimeException(e);
+            }
+         }
+      }
+   }
+
+   protected static <TClass, TAnnotation extends Annotation> void scanFile(String resource, int prefixLength,
+         Class<TClass> superClass, Class<TAnnotation> annotationClass, String requirePackage, Consumer<Class<? extends TClass>> consumer) {
+      if (resource.endsWith(".jar")) {
+         scanJar(resource, superClass, annotationClass, requirePackage, consumer);
+      } else if (resource.endsWith(".class")) {
+         String className = resource.substring(prefixLength, resource.length() - 6).replaceAll("/", ".");
+         scanClass(className, superClass, annotationClass, requirePackage, consumer);
+      }
+   }
+
+   public static <TClass, TAnnotation extends Annotation> void scanJar(String path,
+         Class<TClass> superClass, Class<TAnnotation> annotationClass, String requirePackage, Consumer<Class<? extends TClass>> consumer) {
+      log.tracef("Looking for @%s %s, loading classes from %s", annotationClass.getSimpleName(), superClass.getSimpleName(), path);
+      try (ZipInputStream inputStream = new ZipInputStream(new FileInputStream(path))) {
+         for(;;) {
+            ZipEntry entry = inputStream.getNextEntry();
+            if (entry == null) break;
+            if (!entry.getName().endsWith(".class")) continue;
+            String className = entry.getName().replace('/', '.').substring(0, entry.getName().length() - 6);
+            scanClass(className, superClass, annotationClass, requirePackage, consumer);
+         }
+      } catch (FileNotFoundException e) {
+         log.error("Cannot load executed JAR file '" + path + "'to find stages.");
+      } catch (IOException e) {
+         log.error("Cannot open/read JAR '" + path + "'");
+      }
+   }
+
+   protected static <TClass, TAnnotation extends Annotation> void scanClass(String className,
+         Class<TClass> superClass, Class<TAnnotation> annotationClass, String requirePackage, Consumer<Class<? extends TClass>> consumer) {
+      if (requirePackage != null && !className.startsWith(requirePackage)) return;
+      Class<?> clazz;
+      try {
+         System.setErr(NULL_PRINT_STREAM); // suppress any error output
+         clazz = Class.forName(className);
+         System.setErr(ERR_PRINT_STREAM);
+      } catch (Throwable t) {
+         /* There are other problems during class loading that could lead to Errors -> ignore them */
```

Comment:
Let's catch CNF & Linkage Error then. Cool, it's a good opportunity to make it right.